# 表达式求值
## 中缀表达式转换为后缀表达式
利用栈可以将中缀表达式转换为后缀表达式，具体步骤如下：
- 从左到右遍历中缀表达式的每个元素
- 操作数：直接输出到后缀表达式
- 运算符：将其与栈顶运算符比较
    - 如果栈为空或栈顶是左括号 (，直接入栈
    - 否则如果当前运算符优先级高于栈顶运算符，直接入栈
    - 否则，弹出栈顶运算符并输出到后缀表达式，直到栈顶运算符优先级低于当前运算符或栈为空，然后将当前运算符入栈
- 左括号：直接入栈
- 右括号：将栈中元素弹出并输出到后缀表达式，直到遇到左括号 (，然后将左括号出栈（不输出）
- 遍历结束后，将栈中剩余的运算符依次弹出并输出到后缀表达式

## 后缀表达式求值
后缀表达式的求值直接利用栈进行计算，步骤如下：
- 扫描表达式：从左到右
- 操作数：压入栈
- 运算符：弹出栈顶的两个元素(操作数)，进行运算，并将结果压入栈
- 遍历结束后，栈中剩下的唯一元素就是表达式的值

In [7]:
# 中缀转后缀
def infix_to_postfix(expression):
    precedence = {'+': 1, '-': 1, '*': 2, '/': 2, '^': 3}
    output = []
    operators = [] # 运算符栈

    for char in expression:
        if char.isdigit(): # 操作数
            output.append(char)
        elif char in precedence: # 运算符
            while (operators and operators[-1] != '(' and
                   precedence[operators[-1]] >= precedence[char]): # 栈不为空且栈顶不是左括号，且当前运算符优先级小于等于栈顶运算符
                output.append(operators.pop()) # 弹出栈顶运算符并输出到后缀表达式
            operators.append(char) # 入栈
        elif char == '(':
            operators.append(char)
        elif char == ')':
            while operators and operators[-1] != '(':
                output.append(operators.pop())
            operators.pop()  # Pop the '('

    while operators:
        output.append(operators.pop())

    return ''.join(output)

# 后缀求值
def evaluate_postfix(expression):
    stack = []

    for char in expression:
        if char.isdigit():
            stack.append(int(char))
        else:
            b = stack.pop()
            a = stack.pop()
            if char == '+':
                stack.append(a + b)
            elif char == '-':
                stack.append(a - b)
            elif char == '*':
                stack.append(a * b)
            elif char == '/':
                stack.append(a // b)  # Use // for integer division
            elif char == '^':
                stack.append(a ** b)

    return stack[0]

# 示例
infix_expression = "3+4*2/(1-5)^2"
postfix_expression = infix_to_postfix(infix_expression)
result = evaluate_postfix(postfix_expression)

print(f"Infix Expression: {infix_expression}")
print(f"Postfix Expression: {postfix_expression}")
print(f"Result: {result}")

Infix Expression: 3+4*2/(1-5)^2
Postfix Expression: 342*15-2^/+
Result: 3


## 示例
示例：中缀表达式 3+4*2/(1-5)^2
### 中缀表达式转后缀表达式
我们按照之前提到的步骤，将中缀表达式 3+4*2/(1-5)^2 转换为后缀表达式。

1. 初始状态：
- 输出：[]
- 栈：[]

2. 遍历字符：
- 3：是数字，直接输出到后缀表达式。
    - 输出：['3']
    - 栈：[]
        - 
- +：是运算符，栈为空，直接入栈。
    - 输出：['3']
    - 栈：['+']
        - 
- 4：是数字，直接输出到后缀表达式。
    - 输出：['3', '4']
    - 栈：['+']

- *：是运算符，栈顶是 +，优先级低于 *，直接入栈。
    - 输出：['3', '4']
    - 栈：['+', '*']

- 2：是数字，直接输出到后缀表达式。
    - 输出：['3', '4', '2']
    - 栈：['+', '*']

- /：是运算符，栈顶是 *，优先级相同，弹出 * 并输出，然后 / 入栈。
    - 输出：['3', '4', '2', '*']
    - 栈：['+', '/']

- (：是左括号，直接入栈。
    - 输出：['3', '4', '2', '*']
    - 栈：['+', '/', '(']

- 1：是数字，直接输出到后缀表达式。
    - 输出：['3', '4', '2', '*', '1']
    - 栈：['+', '/', '(']

- -：是运算符，栈顶是 (，直接入栈。
    - 输出：['3', '4', '2', '*', '1']
    - 栈：['+', '/', '(', '-']

- 5：是数字，直接输出到后缀表达式。
    - 输出：['3', '4', '2', '*', '1', '5']
    - 栈：['+', '/', '(', '-']

- )：是右括号，弹出栈顶运算符直到遇到 (，并将 ( 出栈。
    - 输出：['3', '4', '2', '*', '1', '5', '-']
    - 栈：['+', '/']

- ^：是运算符，栈顶是 /，优先级低于 ^，直接入栈。
    - 输出：['3', '4', '2', '*', '1', '5', '-']
    - 栈：['+', '/', '^']

- 2：是数字，直接输出到后缀表达式。
    - 输出：['3', '4', '2', '*', '1', '5', '-', '2']
    - 栈：['+', '/', '^']

3. 遍历结束：
- 将栈中剩余的运算符依次弹出并输出到后缀表达式。
    - 输出：['3', '4', '2', '*', '1', '5', '-', '2', '^', '/', '+']
    - 栈：[]

最终得到的后缀表达式为：342*15-2^/+

### 后缀表达式求值
我们按照之前提到的步骤，对后缀表达式 342*15-2^/+ 进行求值。

1. 初始状态：
- 栈：[]

2. 遍历字符：
- 3：是数字，入栈。
    - 栈：[3]

- 4：是数字，入栈。
    - 栈：[3, 4]

- 2：是数字，入栈。
    - 栈：[3, 4, 2]

- *：是运算符，弹出两个操作数 4 和 2，计算 4*2=8，结果入栈。
    - 栈：[3, 8]

- 1：是数字，入栈。
    - 栈：[3, 8, 1]

- 5：是数字，入栈。
    - 栈：[3, 8, 1, 5]

- -：是运算符，弹出两个操作数 1 和 5，计算 1-5=-4，结果入栈。
    - 栈：[3, 8, -4]

- 2：是数字，入栈。
    - 栈：[3, 8, -4, 2]

- ^：是运算符，弹出两个操作数 -4 和 2，计算 (-4)^2=16，结果入栈。
    - 栈：[3, 8, 16]

- /：是运算符，弹出两个操作数 8 和 16，计算 8/16=0.5，结果入栈。
    - 栈：[3, 0.5]

- +：是运算符，弹出两个操作数 3 和 0.5，计算 3+0.5=3.5，结果入栈。
    - 栈：[3.5]

最终栈中剩下的唯一元素就是表达式的值：3.5

总结
通过栈的应用，我们成功地将中缀表达式 3+4*2/(1-5)^2 转换为后缀表达式 342*15-2^/+，并求得其值为 3.5。这个过程展示了栈在处理表达式求值中的强大功能。